crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,434 @@
+                curr = next;
+                next = strstr(curr, separator);
+                if (next != NULL)
+                    next += strlen(separator);
+                else
+                    next = curr + strlen(curr);
+            }
+            if (next - txt < available_len) /* implies here: *next == '\0' */
+                curr = NULL;
+        }
+        if (curr != NULL) {
+            /* split error msg at curr since error data would get too long */
+            if (curr != txt) {
+                tmp = OPENSSL_strndup(txt, curr - txt);
+                ERR_add_error_data(2, separator, tmp);
+                OPENSSL_free(tmp);
+            }
+            put_error(ERR_LIB_CMP, func, err, file, line);
+            txt = curr;
+        } else {
+            ERR_add_error_data(2, separator, txt);
+            txt = next; /* finished */
+        }
+    } while (*txt != '\0');
+}
+
+/* this is similar to ERR_print_errors_cb, but uses the CMP-specific cb type */
+void OSSL_CMP_print_errors_cb(OSSL_cmp_log_cb_t log_fn)
+{
+    unsigned long err;
+    char component[128];
+    char msg[ERR_print_errors_cb_LIMIT];
+    const char *file, *func = NULL, *data;
+    int line, flags;
+
+    if (log_fn == NULL) {
+#ifndef OPENSSL_NO_STDIO
+        BIO *bio_err = BIO_new_fp(stderr, BIO_NOCLOSE);
+
+        ERR_print_errors(bio_err);
+        BIO_free(bio_err);
+#else
+        CMPerr(0, CMP_R_NO_STDIO);
+#endif
+        return;
+    }
+
+    /* TODO add when available: ERR_peek_error_func(&func); */
+    while ((err = ERR_get_error_line_data(&file, &line, &data, &flags)) != 0) {
+        if (!(flags & ERR_TXT_STRING))
+            data = NULL;
+        BIO_snprintf(component, sizeof(component), "OpenSSL:%s:%s",
+                     ERR_lib_error_string(err), func != NULL ? func : "");
+        BIO_snprintf(msg, sizeof(msg), "%s%s%s", ERR_reason_error_string(err),
+                     data == NULL ? "" : " : ", data == NULL ? "" : data);
+        if (log_fn(component, file, line, OSSL_CMP_LOG_ERR, msg) <= 0)
+            break;              /* abort outputting the error report */
+        /* TODO add when available: ERR_peek_error_func(&func); */
+    }
+}
+
+/*
+ * functions manipulating lists of certificates etc.
+ * these functions could be generally useful.
+ */
+
+int ossl_cmp_sk_X509_add1_cert(STACK_OF(X509) *sk, X509 *cert,
+                               int no_dup, int prepend)
+{
+    if (no_dup) {
+        /*
+         * not using sk_X509_set_cmp_func() and sk_X509_find()
+         * because this re-orders the certs on the stack
+         */
+        int i;
+
+        for (i = 0; i < sk_X509_num(sk); i++) {
+            if (X509_cmp(sk_X509_value(sk, i), cert) == 0)
+                return 1;
+        }
+    }
+    if (!sk_X509_insert(sk, cert, prepend ? 0 : -1))
+        return 0;
+    if (!X509_up_ref(cert)) {
+        (void)sk_X509_delete(sk, prepend ? 0 : sk_X509_num(sk) - 1);
```

Comment:
would be more easy to do first up_ref and handle the error of sk_insert then, right?

Reply:
Indeed! Done.